File: python/87. Scramble String.py
import collections
class Solution(object):
    def isScramble(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: bool
        """
        if s1 == s2:
            return True
        if len(s1) != len(s2):
            return False
        temp1 = collections.defaultdict(int)
        temp2 = collections.defaultdict(int)
        for c1, c2 in zip(s1, s2):
            temp1[c1] += 1
            temp2[c2] += 1
        if temp1 != temp2:
            return False
        l1temp = collections.defaultdict(int)
        l2temp = collections.defaultdict(int)
        r2temp = collections.defaultdict(int)
        for i in range(len(s1) - 1):
            l1temp[s1[i]] += 1
            l2temp[s2[i]] += 1
            r2temp[s2[len(s2) - 1 - i]] += 1
            if l1temp == l2temp:
                if self.isScramble(s1[:i+1], s2[:i+1]) and self.isScramble(s1[i+1:], s2[i+1:]):
                    return True
            elif l1temp == r2temp:
                if self.isScramble(s1[:i+1], s2[-(i + 1):]) and self.isScramble(s1[i+1:], s2[:-(i+1)]):
                    return True
        return False
            
            
            
            

File: cpp/51-100/87. Scramble String.cpp
class Solution {
public:
	bool isScramble(string s1, string s2) {
		if (s1.size() != s2.size())
			return false;
		if(s1 == s2)
			return true;
		map<char, int> dict1;
		map<char, int> dict2;
		for (int i = 0; i < s1.size(); i++) {
			dict1[s1[i]]++;
		}
		for (int i = 0; i < s2.size(); ++i) {
			dict2[s2[i]]++;
		}
		for (int i = 0; i < 26; i++) {
			if (dict1['a' + i] != dict2['a' + i])
				return false;
		}
		for(int i = 1; i < s1.size(); i++)
		{
			if(isScramble(s1.substr(0, i), s2.substr(0, i)) && isScramble(s1.substr(i), s2.substr(i)))
				return true;
			if(isScramble(s1.substr(0, i), s2.substr(len - i)) && isScramble(s1.substr(i), substr(0, len - i)))
				return true
		}
	}
};